server/src/main/java/org/elasticsearch/indices/recovery/MultiFileTransfer.java

```diff
@@ -0,0 +1,171 @@
+ * until all chunk requests are sent/responded.
+ */
+public abstract class MultiFileTransfer<Request extends MultiFileTransfer.ChunkRequest, Response> {
+    private final AtomicBoolean done = new AtomicBoolean(false);
+    private final ActionListener<Void> listener;
+    private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+    private final AsyncIOProcessor<FileChunkResponseItem<Response>> processor;
+    private final int maxConcurrentFileChunks;
+    private long fileOffset = 0;
+    private StoreFileMetaData currentFile = null;
+    private final Iterator<StoreFileMetaData> remainingFiles;
+
+    protected MultiFileTransfer(Logger logger, ThreadContext threadContext, ActionListener<Void> listener,
+                                int maxConcurrentFileChunks, List<StoreFileMetaData> files) {
+        this.maxConcurrentFileChunks = maxConcurrentFileChunks;
+        this.listener = ActionListener.wrap(
+            r -> {
+                if (done.compareAndSet(false, true)) {
+                    listener.onResponse(r);
+                }
+            },
+            e -> {
+                if (done.compareAndSet(false, true)) {
+                    listener.onFailure(e);
+                }
+            });
+        this.processor = new AsyncIOProcessor<>(logger, maxConcurrentFileChunks * 2, threadContext) {
+            @Override
+            protected void write(List<Tuple<FileChunkResponseItem<Response>, Consumer<Exception>>> items) {
+                handleItems(items);
+            }
+        };
+        this.remainingFiles = files.iterator();
+    }
+
+    public final void start() {
+        // put an dummy item to start the processor
+        processor.put(new FileChunkResponseItem<>(SequenceNumbers.UNASSIGNED_PRIMARY_TERM, null, null, null), e -> {});
+    }
+
+    private void handleItems(List<Tuple<FileChunkResponseItem<Response>, Consumer<Exception>>> items) {
+        if (done.get()) {
+            return;
+        }
+        try {
+            for (Tuple<FileChunkResponseItem<Response>, Consumer<Exception>> item : items) {
+                final FileChunkResponseItem<Response> resp = item.v1();
+                if (resp.requestSeqId == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                    continue; // not an actual item
+                }
+                requestSeqIdTracker.markSeqNoAsProcessed(resp.requestSeqId);
+                if (resp.failure != null) {
+                    handleError(resp.md, resp.failure);
+                    throw resp.failure;
+                }
+            }
+            while (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getProcessedCheckpoint() < maxConcurrentFileChunks) {
+                if (currentFile == null) {
+                    if (remainingFiles.hasNext()) {
+                        currentFile = remainingFiles.next();
```

Comment:
Should we add a `openFile` call that the subclass uses to open its input and call here? Currently `nextChunkRequest` ignores the file except when it is about to begin on a new file.

Reply:
++